### PR TITLE
[parallel] fix: NPU Hang/Deadlock during DTensor parameter loading in FSDP2

### PR DESCRIPTION
**Pull request by First-Frost-code**
**Describe the bug**
When training VLM/MoE models on certain hardware backends (especially Ascend NPUs with HCCL), the program frequently hangs indefinitely during the weight loading phase. 

**Root Cause**
The original `_dispatch_parameter` uses `dtensor_factory` and then calls `.copy_()` on a flat parameter. When assigning a DTensor view to a flat parameter, PyTorch's `__torch_dispatch__` implicitly triggers a `Redistribute` collective communication. This uncoordinated implicit network communication disrupts the backend streams on NPU, causing permanent deadlocks.

**Proposed Solution**
This PR adds a defensive feature toggle for NPU environments to bypass implicit DTensor network communications:
* Extracts the local physical tensor using `.to_local()`.
* Manually parses the `placements` to perform pure mathematical `.chunk()`, assigning the correct shard based on the local mesh coordinate.
* Performs a pure local physical `.copy_()`.

This "physical bypass" effectively decouples the weight assignment from the distributed communication graph, completely resolving the deadlock while leaving the default GPU execution path untouched.

### CODE REVIEW

**Comment by FoolPlayer** [comment]
Thanks for you PR, can you help show the deadlock case (like model / size / training args), in fsdp2 we use rank0 load and broadcast to other ranks to avoid OOM. In this way OOM will happen only one tensor is larger than the device max memory https://github.com/ByteDance-Seed/VeOmni/blob/187ac879061e62153f6c1d3831ea54c12ec10a32/veomni/distributed/torch_parallelize.py#L566. And I notice that the fixed code is fsdp1 but the title is fsdp2.

**Comment by First-Frost-code** [comment]
Thanks for pointing that out! You brought up excellent points. Let me clarify the context and the root cause:

1. Regarding FSDP1 vs FSDP2:
You are completely right. I put FSDP2 in the title because my YAML config explicitly uses fsdp_mode: fsdp2. I didn't realize that under the hood, the MoE weight dispatching routing in torch_parallelize.py falls back to the FSDP1 logic. That's a great clarification, and I have updated the PR title to FSDP1.

2. The Deadlock Case (How to Reproduce):

Model: Qwen3-VL-30B-A3B-Instruct

Hardware: 16-card Ascend NPU cluster (e.g., 910C with HCCL)

Why the hang happens when EP > 1:
When ep_size is greater than 1, parallel_plan is generated. During rank0_load_and_broadcast_weights, the broadcasted tensor (like gate.weight) enters _dispatch_parameter. When this DTensor view is assigned via .copy_(), PyTorch's __torch_dispatch__ silently attempts an implicit collective Redistribute. Because HCCL streams are highly sensitive, this unplanned implicit communication instantly causes an AclrtSynchronizeStreamWithTimeout deadlock across the cluster.

3. Why CPU Chunking?
By moving the .chunk() operation to the CPU and doing a pure local physical copy, we strictly bypass dtensor_factory and eradicate the implicit network collective, fixing the deadlock.
(You are absolutely correct that rank0_load_and_broadcast yields one tensor at a time. The CPU chunking I added in my latest commit was indeed to address the gemini-code-assist bot's feedback, ensuring we don't cause an NPU memory spike when sharding a single massive expert tensor, but the core of this PR is to solve the HCCL hang).

Thanks again for the rigorous review! Please let me know if any further adjustments are needed.

File: veomni/models/module_utils.py [hunk 1]
```diff
@@ -188,16 +188,48 @@ def _dispatch_parameter(
...
+            target_local = orig_tensor.to_local()
+
+            if target_local.shape == tensor.shape:
+                target_local.copy_(tensor)
+            else:
+                target_local.copy_(tensor.reshape(target_local.shape))
+                
+        else:
+            # Default execution path for GPUs or non-EP scenarios
+            tensor =
```
Comment (FoolPlayer) [inline veomni/models/module_utils.py]: GPU case missed

**Comment by FoolPlayer** [comment]
Sry,  it's my mistake only fsdp2 use `_dispatch_parameter`. And can you help check whether this PR fix your issue https://github.com/ByteDance-Seed/VeOmni/pull/648.

**Comment by First-Frost-code** [comment]
My apologies! 😅 My code editor completely messed up the copy-paste during my last commit, which accidentally truncated the GPU execution path and broke the syntax. I just pulled the branch back and pushed a clean fix to restore the logic. Thanks for catching that!

Thanks for double-checking and confirming that it is indeed the FSDP2 routing. I have updated the PR title back to FSDP2 to keep it accurate.

Regarding PR #648, I have reviewed its code and logic carefully. It does NOT fix my issue, and this PR (#642) is still absolutely necessary. Here is why:

Different Execution Paths: PR #648 explicitly targets and fixes the load_model_weights (all-ranks-read) path. As the author of #648 noted in their PR description: "The rank0_load_and_broadcast_weights path keeps src_data_rank=0 — it legitimately needs scatter since only rank 0 reads."

My Trigger Condition: In my training setup (loading HF format initially), the framework falls back to the broadcast path. My training logs explicitly show: >> Loading model weights from disk on rank0 then broadcasting to other ranks...

The Deadlock Remains: Because PR #648 leaves the rank0_load_and_broadcast_weights path unchanged, my setup still executes the old logic. It still passes the tensor to dtensor_factory, which triggers the implicit Redistribute collective via PyTorch's __torch_dispatch__. This uncoordinated implicit collective is exactly what destroys the HCCL streams and causes the permanent hang on Ascend NPUs.

Conclusion: PR #648 is a fantastic fix for the direct-read scenario, but this PR (#642) acts as the essential safety net (physical bypass) for the broadcast scenario on NPUs. They complement each other perfectly!

Let me know if you need any more logs or tests from my side!
